server/src/main/java/org/spine3/server/aggregate/Aggregate.java

```diff
@@ -390,7 +413,30 @@ void restore(Snapshot snapshot) {
             builder.mergeFrom(stateToRestore);
             initVersion(versionFromSnapshot);
         } else {
+            injectState(stateToRestore, versionFromSnapshot);
+        }
+    }
+
+    /**
+     * Sets the passed state and version.
+     *
+     * <p>The method circumvents the protection in the {@link #setState(Message, Version)
+     * setState()} method by creating a fake builder instance, which is cleared
+     * after the call.
+     *
+     * <p>This method has package-private access to be accessible by the
+     * {@code AggregateBuilder} test utility class from the {@code testutil} module.
+     */
+    @VisibleForTesting
+    void injectState(S stateToRestore, Version versionFromSnapshot) {
+        try {
+            @SuppressWarnings("unchecked")
+                // The cast is safe as we checked the type on the construction.
+            final B fakeBuilder = (B) getState().newBuilderForType();
+            this.builder = fakeBuilder;
             setState(stateToRestore, versionFromSnapshot);
```

Comment:
`injectState()` is called only once from the `restore(Snapshot)` method. We should have a valid `Snapshot` at this stage. That's why we “inject”.

But I agree that we need yet more protection of an `Aggregate` as we discussed. So, I wanted to discuss having default validation in the `AbstractVersionableEntity.validate()`, which currently does nothing. That would solve the problem of passing invalid state to entities or building an invalid state in an aggregate applier.